### PR TITLE
Business addons

### PR DESCRIPTION
**Pull request by drewsetski**
## Description

In order to link the activation of plans and add-on and unlock it's features `koding button`, `chatlio`, `intercom`, I have to use Redux/Flux later for storing general states.
Deactivations of plan, add-on are not implemented. Also tooltip isn't done, because `Tabs` in `HomeAppView` doesn't except any other prop.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by sinan** [comment]
@drewsetski can you pls share a screencast or some screenshots?

**Comment by drewsetski** [comment]
Here is video:
https://www.dropbox.com/s/nwqdjxpxwgnxgp0/AddOns.mov?dl=0
Interactivity will be finished with Redux states

File: client/component-lab/Banner/Banner.story.coffee [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+          <a href=""><Button type="primary-1" size="medium">EXPLORE</Button></a>
+        </Footer>
+      </Container>
+    </Banner>
+
+
+  .add 'Business Add-On', ->
+
+    <Banner
+      type="primary"
```
Comment (hakankaradis) [inline client/component-lab/Banner/Banner.story.coffee]: https://github.com/koding/styleguide-coffeescript#use-single-quotes-if-there-is-no-string-interpolation

File: client/component-lab/Banner/StoryStyles.stylus [hunk 2]
```diff
@@ -0,0 +1,18 @@
+.BusinessAddOnBanner
+  margin-bottom                     30px
+  padding					  		48px
```
Comment (hakankaradis) [inline client/component-lab/Banner/StoryStyles.stylus]: indentation

File: client/component-lab/Modal/Modal.coffee [hunk 3]
```diff
@@ -55,13 +61,17 @@ exports.Content = Modal.Content = ({ children }) ->
...
+    secondaryButtonType, secondaryButtonTitle, onSecondaryButtonClick, className, size } = props
+  className = [
+    styles.footer
+    className
+  ].join ' '
+  <div className={className}>
     <div className={styles.footerContainer}>
-      <Button type={secondaryButtonType} size='small' onClick={onSecondaryButtonClick}>
+      <Button type={secondaryButtonType} size={'small' if not size} onClick={onSecondaryButtonClick}>
         {secondaryButtonTitle}
```
Comment (hakankaradis) [inline client/component-lab/Modal/Modal.coffee]: Will not be `size='undefined'`  if size ?

File: client/home/lib/billing/components/subscriptionsection.coffee [hunk 4]
```diff
@@ -19,6 +20,7 @@ module.exports = class SubscriptionSection extends React.Component
 
     <DashboardSection title='Koding Subscription'>
       <Subscription {...subscriptionProps} />
+      <AddonsBanner />
```
Comment (hakankaradis) [inline client/home/lib/billing/components/subscriptionsection.coffee]: I believe showing AddonsBanner should be tied to any condition to show it to user.
We might provide you more info about when it should be showed or not

File: client/home/lib/addons/index.coffee [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      if (Object.keys query).length
+        pane.mainView?.handleQuery? query
+
+      break
+
+
+  getActivePlan: ->
+
+    # Do it with Redux/Flux
+    return null
```
Comment (hakankaradis) [inline client/home/lib/addons/index.coffee]: Is there any info that you are waiting from us to implement the Redux stores?

File: client/home/lib/addons/homebusinessaddondeactivation.coffee [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+  constructor: (options = {}, data) ->
+
+    super options, data
+
+  deactivateBusinessAddOn: () ->
+
+    console.log('Business Add-On deactivation')
+
+  renderReact: -> <PlanDeactivation.Container target="BUSINESS ADD-ON" onDeactivation={@bound 'deactivateBusinessAddOn'}/>
```
Comment (hakankaradis) [inline client/home/lib/addons/homebusinessaddondeactivation.coffee]: should be in next line.

File: client/home/lib/addons/homesupportplandeactivation.coffee [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+  getActivePlan: ->
+
+    return 'basic'
+
+  deactivateSupportPlan: () ->
+
+    console.log('Support Plan deactivation')
+
+  renderReact: -> <PlanDeactivation.Container target="#{@getActivePlan()} SUPPORT PLAN" onDeactivation={@bound 'deactivateSupportPlan'}/>
```
Comment (hakankaradis) [inline client/home/lib/addons/homesupportplandeactivation.coffee]: same here

**Review by hakankaradis — CHANGES_REQUESTED** [review]
There are many indentation problems in stylus files and wrong usage of double quote

File: client/component-lab/SupportPlan/SupportPlan.story.coffee [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+      period : 'month'
+      features : [
+        '4 Hours dedicated support'
+        'Stack script support'
+        '24 Hours response time'
+        'General Troubleshooting'
+      ]
+    }
+    <SupportPlan
+          name={plan.name}
```
Comment (hakankaradis) [inline client/component-lab/SupportPlan/SupportPlan.story.coffee]: indentation

File: client/component-lab/Banner/StoryStyles.stylus (hunk 2 above)
Comment (drewsetski) [inline client/component-lab/Banner/StoryStyles.stylus]: perhaps switch between tabs and spaces was jumping back and forth after each opening of file

File: client/home/lib/addons/components/addondeactivation/view.coffee [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+  mapDispatchToProps = (dispatch) ->
+
+    return {
+      onDeactivateBusinessAddOn: ->
+        dispatch(addon.remove())
+    }
+
+
+  module.exports = connect(
```
Comment (hakankaradis) [inline client/home/lib/addons/components/addondeactivation/view.coffee]: view.coffee files should only be responsible for rendering view with given props.
This usage is against the terminology.

File: client/home/lib/addons/components/supportplans/view.coffee [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+    return {
+      onActivateSupportPlan: (plan) ->
+        dispatch(supportplans.create plan)
+      onUpdateSupportPlan: (plan) ->
+        dispatch(supportplans.update plan)
+    }
+
+
+  module.exports = connect(
```
Comment (hakankaradis) [inline client/home/lib/addons/components/supportplans/view.coffee]: same here, these props should be passed from `container`

File: client/home/lib/addons/components/supportplandeactivation/view.coffee [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+  mapDispatchToProps = (dispatch) ->
+
+    return {
+      onDeactivateSupportPlan: ->
+        dispatch(supportplans.remove())
+    }
+
+
+  module.exports = connect(
```
Comment (hakankaradis) [inline client/home/lib/addons/components/supportplandeactivation/view.coffee]: same here!

**Review by hakankaradis — CHANGES_REQUESTED** [review]
Thanks for changes @drewsetski 

There is a problem with usage of Redux.
Please check [this link](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0#.8j9784w0z) to understand the difference between `Presentational and Container Components`

Here is what you need to do.
You will need 2 different containers to manage component's state and passing actions via props.

You need to add a containers folder under `home/lib/addons/` directory
one will be businessaddons.coffee

```coffee
React = require 'react'
{ connect } = require 'react-redux'
addon = require 'app/redux/modules/payment/addon'
BusinessAddons = require '../components/businessaddons'

mapStateToProps = (state) ->
  return {
    addonStatus: addon.isActivated state
    kodingButtonStatus: addon.isKodingButtonActivated state
    intercomStatus: addon.isIntercomActivated state
  }

mapDispatchToProps = (dispatch) ->
  return {
    onActivateBusinessAddOn: -> dispatch(addon.create())
    onDeactivateBusinessAddOn: -> dispatch(addon.remove())
    toggleModal: -> dispatch(addon.toggleModal())
    onActiveKodingButton: -> dispatch(addon.activeKodingButton())
    onActiveIntercom: -> dispatch(addon.activeIntercom())
    onActiveChatlio: -> dispatch(addon.activeChatlio())
  }

module.exports = connect(
  mapStateToProps
  mapDispatchToProps
)(BusinessAddons)
```
And also you will need to add wrapper component to manage e,g, business addons components

You need to pass components states and actions to businessaddons component.

So you will also need to add another folder which is called businesscomponent under `home/lib/addons/components/businessaddons`

`container.coffee` will be like 
```coffee
React = require 'app/react'
View = require './view'
kd = require 'kd'
{ Provider } = require 'react-redux'

module.exports = class BusinessAddOnContainer extends React.Component

  render: ->

    <Provider store={kd.singletons.store}>
      <View />
    </Provider>
```

view.coffee will be like 
```coffee
React = require 'app/react'

Box = require 'lab/Box'
Button = require 'lab/Button'
SupportPlanActivationModal = require 'lab/SupportPlanActivationModal'
{
  Container
  Header
  Message
  Divider
  List
  Footer
  Actions
  PriceSegment
} = Banner = require 'lab/Banner'

class BusinessAddOns extends React.Component

  render: ->

    <div>
      {
        <div>
          <Banner
            type='primary'
            image='banner_business_add_on'
            className='Banner'>
            <Container>
              <Header
                heading='Empower Your Koding with'
                target='Business Add-On'/>
              <Message>
                Fusce dapibus, tellus ac cursus commodo, tortor mauriscondimentum nibh, ut fermentum massa justo sit amet risus.
              </Message>
              <Divider />
              <List
                title='Features'
                items={[
                  'Try on Koding functionality'
                  'Direct communication with leads through <br/> Intercom and Chatlio Integration'
                  'Analytics (Coming soon)'
                ]} />
              <Footer>
                <a href='/Home/add-ons/business-add-on#koding-button'>See what you will get with Business Add-On below</a>
              </Footer>
            </Container>
            <Actions className='Actions'>
              <PriceSegment
                price={@price}
                onClick={@props.toggleActivationModal()}
                link='https://www.koding.com/pricing'/>
            </Actions>
          </Banner>
          <SupportPlanActivationModal
            isOpen={@props.modalOpen}
            title='Business Add On'
            image='business_add_on_activation'
            label='Business Add On'
            price={@price}
            shouldCloseOnOverlayClick={yes}
            onCancel={@props.toggleActivationModal()}
            onActivateSupportPlanClick={@props.activateBusinessAddOn()} />
        </div>  unless @props.addonStatus
      }
    </div>

module.exports = BusinessAddOns
```

You need to do this for support plan tab as well.
I mean you will add another container `home/lib/addons/containers/supportplan.coffee` and also a wrapper component to manage the support plan component's state and its actions

/cc: @usirin can you also please review the code?

**Comment by hakankaradis** [comment]
@sinan Can you please provide us correct copies for those pages.

File: client/component-lab/Banner/Banner.coffee [hunk 12]
```diff
@@ -0,0 +1,114 @@
+React = require 'react'
+Box = require 'lab/Box'
+Button = require 'lab/Button'
+Label = require 'lab/Text/Label'
+generateClassName = require 'classnames'
+formatMoney = require 'app/util/formatMoney'
```
Comment (hakankaradis) [inline client/component-lab/Banner/Banner.coffee]: You are missing that util `formatMoney`

**Review by hakankaradis — CHANGES_REQUESTED** [review]

**Review by hakankaradis — APPROVED** [review]

**Review by hakankaradis — CHANGES_REQUESTED** [review]
@drewsetski 
Everything looks good to me.


But we are missing this feature in this pr.
The selected add-ons and support plans should be visible on the `Team Billing` section.
Once we have this then we can merge this pr

**Comment by hakankaradis** [comment]
Closing this pr, it is reopened here #10425

**Pull request closed without merging**